### PR TITLE
Fix canvas particle material for old GLSL versions

### LINKED ISSUE #25567: CPUParticles2D in gles2 gives white screen when particles_animation in material is checked
**Godot version:**

869887641f3d31f19837a34da116039a26618cb0

**OS/device including version:**

Ubuntu 18.04.1 LTS 64bit 
GeForce GT 540M/PCIe/SSE2
driver : 390.77

**Issue description:**

Screen turns to white if "particles animation" is check in the material of the CPUParticle2D node is checked.
In gles3 is working as expected
**Steps to reproduce:**
^^^

**Minimal reproduction project:**

[gles2_bug.zip](https://github.com/godotengine/godot/files/2824205/gles2_bug.zip)

### PR DESCRIPTION
**Pull request by JFonS**
Attempt to fix #25567. The error only happens on older GLSL versions, so I can't reproduce it myself, but it should be fixed now. Needs testing.

### CODE REVIEW

File: scene/2d/canvas_item.cpp [hunk 1]
```diff
@@ -131,18 +131,11 @@ void CanvasItemMaterial::_update_shader() {
...
-		code += "\t\tframe = clamp(frame, 0, total_frames - 1);\n";
-		code += "\t}\n";
-
-		code += "\tfloat frame_w = 1.0 / h_frames;\n";
-		code += "\tfloat frame_h = 1.0 / v_frames;\n";
-		code += "\tUV.x = UV.x * frame_w + frame_w * float(frame % particles_anim_h_frames);\n";
-		code += "\tUV.y = UV.y * frame_h + frame_h * float(frame / particles_anim_h_frames);\n";
+		code += "\tfloat particle_total_frames = float(particles_anim_h_frames * particles_anim_v_frames);\n";
+		code += "\tfloat particle_frame = floor(INSTANCE_CUSTOM.z * float(particle_total_frames));\n";
+		code += "\tif (!particles_anim_loop) particle_frame=clamp(particle_frame,0.0,particle_total_frames-1.0); else particle_frame=mod(particle_frame,particle_total_frames);\n";
```
Comment (akien-mga) [inline scene/2d/canvas_item.cpp]: There's no `clang-format` running on that inlined code, but try to follow the same style anyway for readability's sake.

**Comment by JFonS** [comment]
It should be better now.

File: scene/2d/canvas_item.cpp [hunk 2]
```diff
@@ -131,19 +131,15 @@ void CanvasItemMaterial::_update_shader() {
...
-
-		code += "\tfloat frame_w = 1.0 / h_frames;\n";
-		code += "\tfloat frame_h = 1.0 / v_frames;\n";
-		code += "\tUV.x = UV.x * frame_w + frame_w * float(frame % particles_anim_h_frames);\n";
-		code += "\tUV.y = UV.y * frame_h + frame_h * float(frame / particles_anim_h_frames);\n";
-
+		code += "\t\tparticle_frame = mod(particle_frame, particle_total_frames);\n";
+		code += "\t}";
+		code += "\tUV /= vec2(h_frames, v_frames);\n";
+		code += "\tUV += vec2(mod(particle_frame, h_frames)/h_frames, floor(particle_frame/h_frames)/v_frames);\n";
```
Comment (akien-mga) [inline scene/2d/canvas_item.cpp]: Also needs spaces around `/` here.

File: scene/resources/material.cpp [hunk 3]
```diff
@@ -617,11 +617,17 @@ void SpatialMaterial::_update_shader() {
...
-			code += "\tif (!particles_anim_loop) particle_frame=clamp(particle_frame,0.0,particle_total_frames-1.0); else particle_frame=mod(particle_frame,float(particle_total_frames));\n";
-			code += "\tUV /= vec2(float(particles_anim_h_frames),float(particles_anim_v_frames));\n";
-			code += "\tUV += vec2(mod(particle_frame,float(particles_anim_h_frames)) / float(particles_anim_h_frames), floor(particle_frame / float(particles_anim_h_frames)) / float(particles_anim_v_frames));\n";
+			code += "\tif (!particles_anim_loop) {\n";
+			code += "\t\tparticle_frame = clamp(particle_frame, 0.0, particle_total_frames - 1.0);\n";
+			code += "\t} else {\n";
+			code += "\t\tparticle_frame = mod(particle_frame, particle_total_frames);\n";
+			code += "\t}";
+			code += "\tUV /= vec2(h_frames, v_frames);\n";
+			code += "\tUV += vec2(mod(particle_frame,h_frames)/h_frames, floor(particle_frame/h_frames)/v_frames);\n";
```
Comment (akien-mga) [inline scene/resources/material.cpp]: And here (as well as after the first comma).

File: scene/2d/canvas_item.cpp (hunk 2 above)
Comment (JFonS) [reply scene/2d/canvas_item.cpp]: I thought spaces were not needed and removed them... Changing it.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**